File: map-reduce/src/main/java/by/bsuir/labs/akka/actor/MapActor.java
package by.bsuir.labs.akka.actor;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

import akka.actor.ActorRef;
import akka.actor.UntypedActor;
import by.bsuir.labs.akka.message.MapData;
import by.bsuir.labs.akka.message.WordCount;

public class MapActor extends UntypedActor {

    private ActorRef reduceActor = null;

    String[] STOP_WORDS = {"a", "am", "an", "and", "are", "as", "at", "be", "do", "go", "if", "in", "is", "it", "of",
            "on", "the", "to"};

    private List<String> STOP_WORDS_LIST = Arrays.asList(STOP_WORDS);

    public MapActor(ActorRef reduceActorRef) {
        reduceActor = reduceActorRef;
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof String) {
            String work = (String) message;
            MapData data = evaluateExpression(work);
            reduceActor.tell(data);
        } else
            unhandled(message);
    }

    private MapData evaluateExpression(String line) {
        List<WordCount> dataList = new ArrayList<WordCount>();
        StringTokenizer parser = new StringTokenizer(line);
        while (parser.hasMoreTokens()) {
            String word = parser.nextToken().toLowerCase();
            if (!STOP_WORDS_LIST.contains(word)) {
                dataList.add(new WordCount(word, 1));
            }
        }
        return new MapData(dataList);
    }
}

File: map-reduce/src/main/scala/by/bsuir/labs/akka/scala/actors/MapActor.scala
package by.bsuir.labs.akka.scala.actors

import java.util.ArrayList
import java.util.StringTokenizer

import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorRef
import by.bsuir.labs.akka.scala.{MapData, Word}

class MapActor(reduceActor: ActorRef) extends Actor {

  val STOP_WORDS_LIST = List("a", "am", "an", "and", "are", "as", "at", "be", "do", "go", "if", "in", "is", "it", "of", "on", "the", "to")

  def receive: Receive = {
    case message: String =>
      reduceActor ! evaluateExpression(message)
    case _ =>

  }

  def evaluateExpression(line: String): MapData = {
    val dataList = new ArrayList[Word]
    val parser: StringTokenizer = new StringTokenizer(line)
    val defaultCount: Integer = 1
    while (parser.hasMoreTokens) {
      val word: String = parser.nextToken().toLowerCase
      if (!STOP_WORDS_LIST.contains(word)) {
        dataList.add(new Word(word, defaultCount))
      }
    }
    new MapData(dataList)
  }
}
